Add support for system text scale factor.

gspencergoog opened the pull request:
This adds support for monitoring the textScaleFactor from the platform's system settings.  It hooks into the MediaQuery, so any app descended from WidgetsApp will automatically get support for handling the font changes.

In order for Android to not restart the app when the system font scale changes, I had to add fontScale to the manifests.

Addresses #10524

gspencergoog committed:
Add support for system text scale factor.

gspencergoog committed:
Add support for system text scale factor.

gspencergoog commented:
Actually, hang on for a bit on this review.  I need to add tests (and fix whatever that travis failure is).

gspencergoog commented:
Ahh, the Travis failure is just because it depends upon the engine change.  I'll land that before this one so the checks will pass.

goderbauer commented on examples/flutter_gallery/android/app/src/main/AndroidManifest.xml:
```diff
@@ -21,7 +21,7 @@
                   android:launchMode="singleTop"
                   android:theme="@android:style/Theme.Light.NoTitleBar"
-                  android:configChanges="orientation|keyboardHidden|keyboard|screenSize|locale|layoutDirection"
+                  android:configChanges="orientation|keyboardHidden|keyboard|screenSize|locale|layoutDirection|fontScale|fontScale"
```
remove one fontScale?

gspencergoog replied:
Whoops.  Done.

goderbauer reviewed (commented):
Can you add some tests?

gspencergoog committed:
Review changes.

gspencergoog commented on packages/flutter/lib/src/widgets/app.dart:
```diff
@@ -399,6 +399,14 @@ class _WidgetsAppState extends State<WidgetsApp> implements WidgetsBindingObserv
   @override
+  void didChangeTextScaleFactor() {
+    setState(() {
+      // The properties of ui.window have changed. We use them in our build
```
Done

Hixie commented on packages/flutter/lib/src/widgets/binding.dart:
```diff
@@ -148,6 +148,53 @@ abstract class WidgetsBindingObserver {
+  ///
+  ///   @override
+  ///   Widget build(BuildContext context) {
+  ///     return new Text('Last scale factor: _lastTextScaleFactor');
```
should this be "Current"?

gspencergoog replied:
Yes.  Fixed.

Hixie commented:
A test that verifies that this propagates all the way to the various widgets that have text in them would be great.

gspencergoog committed:
Add test for textScaleFactor propagation

gspencergoog committed:
Review changes

gspencergoog committed:
Making a better test.

Hixie commented on packages/flutter/test/material/app_test.dart:
```diff
@@ -307,16 +307,27 @@ void main() {
+        textScaleFactor = MediaQuery.of(context).textScaleFactor;
+        return new Container();
+      }),
+    ));
```
What I meant was a test that actually verified that the Text of, say, an AppBar widget, or of a FlatButton, actually grew.

gspencergoog replied:
Well, but shouldn't that be tested by a different test?  Seems like as long as the textScaleFactor changes in the media query, then this code has done its job.  Looks like the MediaQuery and Text tests don't really do that kind of testing, however.  I'll see if I can add something there.

Hixie replied:
I don't mind how many tests you write or how you split it up. :-)

Until now we've not done anything with textScaleFactor, so it's completely untested. We should verify that widgets get bigger when the font is scaled. We can test this as one end-to-end test or many tests that check each link in the chain, either way is fine by me.

gspencergoog replied:
So, I tried adding a test to text_test.dart, and I'm perplexed.  Why does this fail at the expect for `largeSize.height`?  It thinks that text with scale factor 3.0 is the same size as text with a scale of 1.0 (14 px).

```dart
  testWidgets('Text respects textScaleFactor', (WidgetTester tester) async {
    await tester.pumpWidget(const Center(
        child: const Text('Hello', textScaleFactor: 3.0, textDirection: TextDirection.ltr)
    ));

    final RichText text = tester.firstWidget(find.byType(RichText));
    expect(text.textScaleFactor, 3.0);
    final Size largeSize = tester.getSize(find.byType(RichText));
    expect(largeSize.height, equals(3.0 * 14.0));
  });
```
Is this some issue with the Linux engine vs the regular engine?  Because text is clearly scaled on a device.  I just can't see how paragraph layout could work on the tests if it can't reliably determine the text size.

Note that this doesn't even involve MediaQuery, so it doesn't have anything to do with that.

gspencergoog replied:
Turns out that this is a bug in the way the default text size is handled. I fixed it and added some more tests.

gspencergoog committed:
Added several tests, fixed a bug in paragraph rendering

gspencergoog committed:
Added testing of text span scaling

Hixie commented on packages/flutter/lib/src/painting/text_style.dart:
```diff
@@ -247,6 +247,9 @@ class TextStyle extends Diagnosticable {
   /// its size.
   final double fontSize;
```
Please document that the fontSize defaults to 14 at the paragraph level.

Hixie commented on packages/flutter/test/widgets/text_test.dart:
```diff
@@ -25,13 +25,58 @@ void main() {
+        child: const Text('Hello',
+            style: const TextStyle(fontSize: 20.0),
+            textScaleFactor: 3.0,
+            textDirection: TextDirection.ltr)
```
please wrap using flutter style

Hixie commented on packages/flutter/test/widgets/text_test.dart:
```diff
@@ -25,13 +25,58 @@ void main() {
+
+  testWidgets('Text respects textScaleFactor with explicit font size',
+      (WidgetTester tester) async {
+    await tester.pumpWidget(const Center(
```
This indentation is confusing (the less indented line is inside the thing declared by the more indented line).
I recommend just not wrapping here, but if you want to wrap then I'd use two rather than four spaces.

Hixie commented on packages/flutter/test/widgets/text_test.dart:
```diff
@@ -25,13 +25,58 @@ void main() {
+  testWidgets('Text respects textScaleFactor with default font size',
+      (WidgetTester tester) async {
+    await tester.pumpWidget(
+        const Center(child: const Text('Hello', textDirection: TextDirection.ltr)));
```
use flutter style (if there's a line break after a (, there should be a line break before the ); also, two-space indent

Hixie commented on packages/flutter/test/rendering/paragraph_test.dart:
```diff
@@ -222,6 +222,51 @@ void main() {
+          new TextSelection(baseOffset: i, extentOffset: i + 1)));
+    }
+    expect(boxes.length, equals(4));
+    expect(boxes[0].toRect().width, equals(25.0));
```
you could just do
```dart
  expect(boxes[0].toRect(), const Rect(25.0, 25.0));
```

Hixie commented on packages/flutter/test/rendering/paragraph_test.dart:
```diff
@@ -222,6 +222,51 @@ void main() {
+    // Test the sizes of nested spans.
+    final List<ui.TextBox> boxes = <ui.TextBox>[];
+    final String text = testSpan.toStringDeep();
+    for (int i = 0; i < text.length; ++i) {
```
(I personally prefer `i += 1`, but that's not a Flutter style issue. Feel free to leave it as `++i`.)

Hixie replied:
I learnt from Pascal, where it's spelt `Inc(Index);`. :-)

I'm averse to `++i` because it looks like an expression but is a statement with an effect. In general I prefer expressions to be side-effect free, because side-effects are a bug factory.

Hixie commented on packages/flutter/test/rendering/paragraph_test.dart:
```diff
@@ -222,6 +222,51 @@ void main() {
+    final String text = testSpan.toStringDeep();
+    for (int i = 0; i < text.length; ++i) {
+      boxes.addAll(paragraph.getBoxesForSelection(
+          new TextSelection(baseOffset: i, extentOffset: i + 1)));
```
nit: wrapping

Hixie commented on packages/flutter/test/rendering/paragraph_test.dart:
```diff
@@ -222,6 +222,51 @@ void main() {
+      ],
+    );
+    final RenderParagraph paragraph = new RenderParagraph(testSpan,
+        textDirection: TextDirection.ltr, textScaleFactor: 2.5);
```
nit: prefer

```dart
    final RenderParagraph paragraph = new RenderParagraph(
      testSpan,
      textDirection: TextDirection.ltr,
      textScaleFactor: 2.5,
    );
```
...so that things with the same semantic level are vertically aligned.

Hixie commented:
I think you should test some of the other widgets too, e.g. make sure that AppBar grows appropriately, ListTile grows appropriately, etc.

gspencergoog committed:
Removed TestingHooks

gspencergoog committed:
Review Changes

gspencergoog committed:
More review changes.

gspencergoog committed:
Adding more tests

gspencergoog committed:
Adding FlatButton test

gspencergoog commented:
OK, so after investigating AppBar and ListTile, both of them have hard-coded sizes.  AppBar uses the constant kToolbarHeight (56 dp), and ListTile just has [hard-coded values](https://github.com/flutter/flutter/blob/1a3097025dbc13969c6ada32cc559023984fdda1/packages/flutter/lib/src/material/list_tile.dart#L380) based on whether the theme is "dense" or not.

I've asked for clarification from the Material Design team about what should happen in cases where the spec says a widget has a fixed size, but the widget contains text.  I'd like to implement any fallout changes from that discussion in another CL.

Added a test for FlatButton. Although it has a fixed height, it does scale in width.

Hixie commented on packages/flutter/lib/src/widgets/app.dart:
```diff
@@ -486,7 +495,7 @@ class _WidgetsAppState extends State<WidgetsApp> implements WidgetsBindingObserv
       }
       return true;
-    }());
+    });
```
mismerge

gspencergoog replied:
Fixed.

Hixie commented on packages/flutter/lib/src/widgets/binding.dart:
````diff
@@ -148,6 +148,53 @@ abstract class WidgetsBindingObserver {
+  /// preferences, and it should affect all of the text sizes in the
+  /// application.
+  ///
+  /// ```dart
````
this should definitely be marked `## Sample code`, to ensure it's properly analyzed

gspencergoog replied:
Cool, I didn't know about the "## Sample code" thing.  Done.

Hixie commented on packages/flutter/lib/src/widgets/binding.dart:
```diff
@@ -297,6 +344,16 @@ abstract class WidgetsBinding extends BindingBase with GestureBinding, RendererB
+  /// Called when the platform text scale factor changes.
+  ///
+  /// See [Window.onTextScaleFactorChanged].
+  @override
```
the docs should be inherited, you shouldn't need to doc it here again

gspencergoog replied:
Yeah, I thought so, but was following handleMetricsChanged pattern.  Removed here, and on handleMetricsChanged, above.

Hixie commented on packages/flutter/lib/src/widgets/media_query.dart:
```diff
@@ -113,7 +113,9 @@ class MediaQueryData {
 
   @override
-  String toString() => '$runtimeType(size: $size, devicePixelRatio: $devicePixelRatio, textScaleFactor: $textScaleFactor, padding: $padding)';
+  String toString() => '$runtimeType(size: $size, '
```
nit: avoid => if you have a line break in the expression

gspencergoog replied:
Done.

Hixie commented on packages/flutter/test/material/buttons_test.dart:
```diff
@@ -44,6 +44,46 @@ void main() {
+            data: new MediaQueryData(textScaleFactor: 1.0),
+            child: new Center(
+              child: new FlatButton(
+                  onPressed: () {},
```
the most trivial nit ever: we use `{ }` for empty code blocks and `{}` for empty maps

gspencergoog replied:
Oooh, tricky.  Done.

Hixie commented on packages/flutter/test/material/buttons_test.dart:
```diff
@@ -44,6 +44,46 @@ void main() {
+            child: new Center(
+              child: new FlatButton(
+                  onPressed: () {},
+                  child: const Text('ABC')
```
trailing comma

gspencergoog replied:
Fixed.

Hixie commented on packages/flutter/test/material/buttons_test.dart:
```diff
@@ -44,6 +44,46 @@ void main() {
+              ),
+            ),
+          ),
+        ),
```
ditto ditto for this test

gspencergoog replied:
Fixed**3.

Hixie commented on packages/flutter/test/material/buttons_test.dart:
```diff
@@ -44,6 +44,46 @@ void main() {
+            child: new Center(
+              child: new FlatButton(
+                  onPressed: () {},
+                  child: const Text('ABC')
```
also your indentation is 2 off in the inner block

gspencergoog replied:
Fixed.

Hixie commented:
![LGTM, but travis is angry.](https://cloud.githubusercontent.com/assets/551196/16323131/36570fa8-395d-11e6-9822-7a0aac54e3b5.png)

goderbauer reviewed (approved):
LGTM when Travis is happy

gspencergoog commented:
Travis won't be happy until the engine change is in and I roll the engine.

Hixie commented:
You can do the roll as part of this patch, that's what I usually do when making a roll-dependent change. It makes Travis happy and gives you more real verification. (Then if it takes too long to land the PR and the engine has rolled on trunk, just remove that part of the PR so you don't roll backwards!)

gspencergoog commented:
Which hash will I put in the engine roll? (e.g. which hash does git squash the engine change to when I squash and merge?)  The last one on the engine change?

gspencergoog committed:
Review changes

Hixie commented:
If you look at the wiki page for releasing the engine, it has a link to a page on GitHub that lists all the hashes.

gspencergoog commented:
Ahh, so I have to wait until the engine change is committed so I have the right hash.  OK.

gspencergoog committed:
Removed a space in comment.

gspencergoog committed:
Rolling engine to 14aab33edd199fe453cd01b9c8a05e9dc3c6679f

gspencergoog committed:
Rolling engine to 14aab33edd199fe453cd01b9c8a05e9dc3c6679f

gspencergoog committed:
Fixing roundoff in test.

gspencergoog committed:
Updating text tests.

gspencergoog committed:
Fix analysis warning in sample code.

gspencergoog committed:
Resetting test scales to avoid kerning changes.

gspencergoog committed:
Fixing roundoff error again? Hmm.

Hixie commented on packages/flutter/test/widgets/text_test.dart:
```diff
@@ -47,7 +47,7 @@ void main() {
     expect(text.textScaleFactor, 3.0);
     final Size largeSize = tester.getSize(find.byType(RichText));
-    expect(largeSize.width, equals(211.0));
+    expect(largeSize.width, equals(210.0));
```
This should be 210 exactly (5 characters * 14 pixels font size * 3 scale factor = 210).
In the tests we use a font that has ASCII characters that are exactly the font size high and exactly the font size wide, (with their baseline at exactly 2/10th of the way from the bottom).

Hixie commented on packages/flutter/test/rendering/paragraph_test.dart:
```diff
@@ -261,13 +261,13 @@ void main() {
     }
     expect(boxes.length, equals(4));
-    expect(boxes[0].toRect().width, equals(25.0));
+    expect(boxes[0].toRect().width, equals(26.0));
```
this should be 25, because 10 * 2.5 = 25.

Hixie commented on packages/flutter/test/rendering/paragraph_test.dart:
```diff
@@ -261,13 +261,13 @@ void main() {
       ));
```
paragraph.size.width above should be (10+20+20+10)*2.5 = 150 exactly.

Hixie commented on packages/flutter/test/rendering/paragraph_test.dart:
```diff
@@ -261,13 +261,13 @@ void main() {
+    expect(boxes[0].toRect().width, equals(26.0));
     expect(boxes[0].toRect().height, equals(25.0));
-    expect(boxes[1].toRect().width, equals(50.0));
+    expect(boxes[1].toRect().width, equals(51.0));
```
20 * 2.5 = 50

Hixie commented on packages/flutter/test/rendering/paragraph_test.dart:
```diff
@@ -261,13 +261,13 @@ void main() {
     expect(boxes[2].toRect().width, equals(50.0));
     expect(boxes[2].toRect().height, equals(50.0));
-    expect(boxes[3].toRect().width, equals(25.0));
+    expect(boxes[3].toRect().width, equals(26.0));
```
10 * 2.5 = 25

gspencergoog committed:
Resetting test scales to avoid kerning changes.

gspencergoog committed:
Testing to see if failures on Linux disappear

gspencergoog committed:
Tests now work on Linux...Hmm.

gspencergoog commented:
OK, going back in time a week didn't help, so I guess I just didn't run them all on my Mac.  Now I'm stuck with figuring out the difference in the Mac and Linux builds and looking for rendering differences there.  It's possible that since text scaling had virtually no tests or usage before now, that it's just a latent bug.

goderbauer commented:
> ... because it runs out of open files

We have an issue on file for this. The command in the very last comment helped me to work around it: https://github.com/flutter/flutter/issues/2976

gspencergoog committed:
Making tests fuzzy enough to pass on Mac and Linux both.

gspencergoog committed:
Adding comments/todos for text size differences.

gspencergoog commented:
OK, it's not always Linux that is wrong, and sometimes Linux is "right". Filed #12357 and made the tests just fuzzy enough to pass.  It's not worth tracking this down until after libtxt rendering is in.

Hixie commented:
Have you verified that this happens without the text scale factor?

gspencergoog commented:
Yes, I reverted the engine to before my engine change, and removed the textScaleFactor setting from the tests, and the results were still inconsistent, depending upon the font size (some size combinations are consistent, some are not).

gspencergoog committed:
Adding more TODOs for #12357

gspencergoog committed:
Rolling engine to e3404b81a53ba3180c7623a6f2190ebb28518f30
